source/common/network/connection_impl.cc

```diff
@@ -89,6 +89,7 @@ ConnectionImpl::ConnectionImpl(Event::DispatcherImpl& dispatcher, int fd,
       // Set a special error state to ensure asynchronous close to give the owner of the
       // ConnectionImpl a chance to add callbacks and detect the "disconnect"
       state_ |= InternalState::BindError;
+      state_ &= ~InternalState::Connecting;
       // Indicate this bind close should "flush", i.e. wait for a dispatcher callback cycle.
       close(ConnectionCloseType::FlushWrite);
```

Comment:
Can you try just replacing this line with:
`file_event_->activate(Event::FileReadyType::Write);`

I think that is a better fix? I don't think the connecting bit flip should even be necessary? I don't recall the history of why we called close() here but from looking at code I don't think it's necessary.

Reply:
P.S. I think you might be able to clean up the bind flag checking logic inside close() also.

Reply:
Seems better. In the meantime I'd figured out another way to avoid this, but with a more complicated test.